src/compile/data/optimizers.ts

```diff
@@ -0,0 +1,150 @@
+import {hasIntersection} from '../../util';
+import {AggregateNode} from './aggregate';
+import {BinNode} from './bin';
+import {DataFlowNode, isNewFieldNode} from './dataflow';
+import {FacetNode} from './facet';
+import {ParseNode} from './formatparse';
+import {NullFilterNode} from './nullfilter';
+import {SourceNode} from './source';
+import {StackNode} from './stack';
+import {TimeUnitNode} from './timeunit';
+import {CalculateNode, FilterNode} from './transforms';
+
+/**
+ * Start optimization path at the leaves. Useful for merging up things.
+ */
+export function optimizeFromLeaves(f: (node: DataFlowNode) => void) {
+  function optimizeNextFromLeaves(node: DataFlowNode) {
+    if (!node || !node.parent || node.parent instanceof SourceNode) {
```

Comment:
I’m wondering:
1) Can nodes that are not SourceNode have no parent?
2) Can `SourceNode`s have parent?